### PR TITLE
Update dependencies, workflows, and configurations (mostly client side)

### PR DESCRIPTION
**Pull request by kklamberty**
The focus of this PR is to get the structure and requirements from client side of the Iteration Template into this lab. This is likely to break things. It is unlikely to pass the checks until at least the changes within the Angular files are up to date. I did change the github actions for both client and server here because I was not thinking.

### CODE REVIEW

File: client/src/app/app.module.ts [hunk 1]
```diff
@@ -21,56 +23,52 @@ import { MatSnackBarModule } from '@angular/material/snack-bar';
...
   MatRadioModule,
-  MatSnackBarModule
+  MatSidenavModule,
+  MatSelectModule,
+  MatSnackBarModule,
+  MatToolbarModule,
+  MatTooltipModule,
 ];
 
 @NgModule({
```
Comment (kklamberty) [inline client/src/app/app.module.ts:54]: I also tried to alphabetize elsewhere. @floogulinc suggests (correctly) that order can matter for these, so I'm going to make another commit that puts back the dangerous things into the order they were in before. This didn't seem to cause problems, but it was an arbitrary decision just meant to help readability. If it's not safe to do, better to leave it confusing.

**Comment by kklamberty** [comment]
My change to the `package-lock.json` file seems to have taken out a lot of "resolved" and "integrity" elements, and I'm not sure if that matters. See: [5c90459](https://github.com/UMM-CSci-3601/3601-lab4-mongo/pull/55/commits/5c90459cbaac9682b607e3230175a62d65ab7d0c) If you would like to verify any of that.

File: .github/workflows/client.yaml [hunk 2]
```diff
@@ -12,23 +12,25 @@ jobs:
...
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
 
     - name: Cache Node modules
-      uses: actions/cache@v2
+      uses: actions/cache@v3
       with:
         path: ~/.npm
         key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
```
Comment (NicMcPhee) [inline .github/workflows/client.yaml]: You still have all the old (and I think non-functional) caching, but aren't using the new `npm` caching command. I would definitely fix that as I think it will speed things up for the students.

File: .github/workflows/e2e.yaml [hunk 3]
```diff
@@ -14,18 +14,18 @@ jobs:
...
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
 
     - name: Cache node modules
-      uses: actions/cache@v2
+      uses: actions/cache@v3
       with:
         path: ~/.npm
         key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
```
Comment (NicMcPhee) [inline .github/workflows/e2e.yaml:23]: Same issue with the caching here as well.

File: client/src/app/users/add-user.component.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ import { UserService } from './user.service';
 })
 export class AddUserComponent implements OnInit {
 
-  addUserForm: FormGroup;
+  addUserForm: UntypedFormGroup;
```
Comment (NicMcPhee) [inline client/src/app/users/add-user.component.ts:15]: Thinking about it, we almost certainly want to move to typed form groups since then the type checker can help us out more. We should probably create an issue for that.

**Review by NicMcPhee — CHANGES_REQUESTED** [review]
This looks great, although I do think you want to fix the `npm` caching business in two of the `.github/workflows` files.

Thanks for grinding through this huge amount of work. We really do need to find a better way to manage these complex collections of repositories so we don't repeat this work all the time.

File: .github/workflows/client.yaml (hunk 2 above)
Comment (kklamberty) [reply .github/workflows/client.yaml]: Can you help -- is this a comment on line 21? This matches what we have in the iteration database, and I thought we were using the newest one there.

File: client/src/app/users/add-user.component.ts (hunk 4 above)
Comment (kklamberty) [reply client/src/app/users/add-user.component.ts:15]: I agree; It just happened to not seem broken when it was just listed as being a FormGroup... but I'm not sure if that's meaningful TBH. So, I went ahead and made this a noticeable thing to address.

**Review by NicMcPhee — DISMISSED** [review]

**Comment by jpwalbran** [comment]
This looks like it does what we want it to:
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows
Haven't dug into it yet.

**Review by NicMcPhee — DISMISSED** [review]

**Review by NicMcPhee — DISMISSED** [review]

**Review by NicMcPhee — APPROVED** [review]

File: .github/workflows/e2e.yaml [hunk 5]
```diff
@@ -14,18 +14,18 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
 
     - name: Cache node modules
```
Comment (floogulinc) [inline .github/workflows/e2e.yaml:19]: Now that you're having the caching from `actions/setup-node` you can remove this entire step.

File: client/src/app/users/user-profile.component.html [hunk 6]
```diff
@@ -1,5 +1,5 @@
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
```
Comment (floogulinc) [inline client/src/app/users/user-profile.component.html:2]: Looks like you've left some stuff from Angular flex-layout here. Hasn't that been removed?

File: client/src/app/users/user-list.component.html [hunk 7]
```diff
@@ -59,25 +59,25 @@
   </div>
 </div>
 
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10" *ngIf="serverFilteredUsers; else usersError" >
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10" *ngIf="serverFilteredUsers; else usersError" >
```
Comment (floogulinc) [inline client/src/app/users/user-list.component.html:63]: Same here there are attributes for the old flex layout library that don't do anything still here and probably in many other places.

File: client/src/app/users/user-list.component.html [hunk 8]
```diff
@@ -1,11 +1,11 @@
-<div fxLayout="row">
-  <div fxFlex fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
+<div class="flex-row">
+  <div class="flex-1" fxFlex.gt-sm="80" fxFlexOffset.gt-sm="10">
```
Comment (floogulinc) [inline client/src/app/users/user-list.component.html:2]: See previous 2 comments.

**Comment by floogulinc** [comment]
I'm not going to leave individual comments for each of them but it seems in many places there are still attributes in html elements for the Angular flex-layout library that was removed. Anything starting with `fxFlex` for example won't be doing anything anymore. This was probably missed in the iteration template as well.

File: .github/workflows/e2e.yaml (hunk 5 above)
Comment (floogulinc) [reply .github/workflows/e2e.yaml:19]: @kklamberty did you merge without removing this?

Comment (kklamberty) [reply .github/workflows/e2e.yaml:19]: Yes. I'm not confident that you are right (or wrong) about this step being unneceassary. The node cache step specifically says it does not cache node_modules, and this step aims to do that. We are not sure if it's working or not working, but it seems that it's _different_ in the least. 

This step does not cache node_modules, as is noted here: https://github.com/actions/setup-node
Different sources have varying opinions on whether or not we *should* cache node_modules, but since our step name suggests that we were trying to do so, I didn't think it would be problematic to keep our attempt in here.

In fact - @NicMcPhee and @jpwalbran - I think that looking at what to do if we have a cache hit might do the trick.
https://github.com/actions/cache#skipping-steps-based-on-cache-hit - I don't think we were ever looking for the cache and behaving differently based on whether or not we had a cache hit.

Comment (floogulinc) [reply .github/workflows/e2e.yaml:19]: This step _isn't_ caching the `node_modules` directory (which we shouldn't be). It sets up caching just for `~/.npm` (see the `path` below) which is the same as what setup-node is doing.

Comment (floogulinc) [reply .github/workflows/e2e.yaml:19]: `~/.npm` is a directory npm uses specifically for caching. We should not cache `node_modules` as that wouldn't work with `npm ci` and could break when the node version changes between runs.
